subsys/usb/host/usbh_ch9.c

```diff
@@ -33,139 +45,188 @@ int usbh_req_setup(const struct device *dev,
 		.wLength = sys_cpu_to_le16(wLength),
 	};
 	struct uhc_transfer *xfer;
-	struct net_buf *buf;
 	uint8_t ep = usb_reqtype_is_to_device(&req) ? 0x00 : 0x80;
 	int ret;
 
-	xfer = uhc_xfer_alloc(dev, addr, ep, 0, 64, SETUP_REQ_TIMEOUT, NULL, NULL);
+	xfer = usbh_xfer_alloc(udev, ep, 0, 64, SETUP_REQ_TIMEOUT, (void *)ch9_req_cb);
 	if (!xfer) {
 		return -ENOMEM;
 	}
 
 	memcpy(xfer->setup_pkt, &req, sizeof(req));
 
-	if (wLength) {
-		buf = uhc_xfer_buf_alloc(dev, wLength);
-		if (!buf) {
-			ret = -ENOMEM;
-			goto buf_alloc_err;
-		}
-
-		if (usb_reqtype_is_to_device(&req) && data != NULL) {
-			net_buf_add_mem(buf, data, wLength);
-		}
-
-		ret = uhc_xfer_buf_add(dev, xfer, buf);
+	if (wLength && buf != NULL) {
+		ret = usbh_xfer_buf_add(udev, xfer, buf);
 		if (ret) {
 			goto buf_alloc_err;
 		}
 	}
 
-	return uhc_ep_enqueue(dev, xfer);
+	ret = usbh_xfer_enqueue(udev, xfer);
+	if (ret) {
+		goto buf_alloc_err;
+	}
+
+	k_sem_take(&ch9_req_sync, K_MSEC(100));
```

Comment:
Why 100 ms? `ch9_req_sync` is given after the control transfer finishes or did I skip something important? If yes, then this is lower than what USB 2.0 specification allows. Also blocking here really kills performance because it prohibits other transfers from executing (for one device this is not such a big deal, but also e.g. prevents bulk transfer to execute alongside control transfer).

Reply:
Semaphore should only be used for synchronous transfer, I would like to move it to `struct usb_device{}`, but then additional work is also needed. This is on my list.